Switches to "/" convention in ghcr.io images

potiuk opened the pull request:
We are using ghcr.io as image cache for our CI builds and Breeze
and it seems ghcr.io is being "rebuilt" while running.

We had been using "airflow-<branch>.." image convention before,
bacause multiple nesting levels of images were not supported,
however we experienced errors recently with pushing 2.1 images
(https://issues.apache.org/jira/browse/INFRA-22124) and during
investigation it turned out, that it is possible now to use "/"
in the name of the image, and while it still does not introduce
multiple nesting levels and folder structure, the UI of GitHub
treats it like that and if you have image which starts wiht
"airflow/", the airflow prefix is stripped out and you can also
have even more "/" in then name to introduce further hierarchy.

Since we have to change image naming convention due to (still
unresolved) bug with no permission to push the v2-1-test image
we've decided to change naming convention for all our cache
images to follow this - now available - "/" connvention to make
it better structured and easier to manage/understand.

Some more optimisations are implemented - Python, prod-build and
ci-manifest images are only pushed when "latest" image is prepared.
They are not needed for the COMMIT builds because we only need
final images for those builds. This simplified the code quite
a bit.

CONTINUE_ON_PIP_CHECK_ERROR variable has been removed in favour
of ignoring pip error when installing dependencies from branch
tip. This might potentially happen for a short while when
new changes have been merged, but constraints were not yet
regenerated and we have conflicting dependencies.

The .dockerignore was reviewed and builds were optimized for
people who locally built provider packages and documentation,
by excluding unnecessary files. Some instructions which run after
the COPY . which did not need sourcer were moved before
the COPY command. Those optimisatiions save 30-40 seconds of
overhead when building the image (especially when you build
images incrementally rather than rebuilding from scratch).

PIP and HELM versions have been updated to latest available.

Backwards-compatibility was implemented to allow PRs that have
not been upgraded to continue building after this one is merged,
also a workaround has been implemented to make this change
to work even if it is not merged yet to main.

This "legacy" mode will be removed in ~week when everybody rebase
on top of main.

Documentation is updated reflecting those changes.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb commented on .dockerignore:
```diff
@@ -40,9 +40,6 @@
 !scripts/docker
 
-# Add provider packages to the context
-!provider_packages
```
Did you mean to include this change here?

potiuk replied:
I can easily separate it out. It's also optimizing the image build experience, but I can move it to separate PR.

potiuk replied:
Done.

ashb commented on .github/workflows/ci.yml:
```diff
@@ -1227,10 +1123,8 @@ ${{ hashFiles('.pre-commit-config.yaml') }}"
-      github.ref == 'refs/heads/main' || github.ref == 'refs/heads/v1-10-test' ||
-      github.ref == 'refs/heads/v2-0-test' || github.ref == 'refs/heads/v2-1-test'
+    # Only run it for direct pushes and scheduled builds
+    if: github.event_name == 'push' || github.event_name == 'schedule'
```
This will run if any committer pushes a branch to apache/airflow instead of their own fork -- and I don't think we want that to happen.

potiuk replied:
Not really :). I was looking on how to generalize that, but I noticed that there is the "push" if at the top of the file:

```
  push:
    branches: ['main', 'v[0-9]+-[0-9]+-test']
```

ashb replied:
Ah cool!

ashb reviewed (changes requested):
There are more changes in here than just chaning from `-` to `/` which makes it very hard to review.

potiuk committed:
Switches to "/" convention in ghcr.io images

We are using ghcr.io as image cache for our CI builds and Breeze
and it seems ghcr.io is being "rebuilt" while running.

We had been using "airflow-.." image convention before,
bacause multiple nesting levels of images were not supported,
however we experienced errors recently with pushing 2.1 images
(https://issues.apache.org/jira/browse/INFRA-22124) and during
investigation it turned out, that it is possible now to use "/"
in the name of the image, and while it still does not introduce
multiple nesting levels and folder structure, the UI of GitHub
treats it like that and if you have image which starts wiht
"airflow/", the airflow prefix is stripped out and you can also
have even more "/" in then name to introduce further hierarchy.

Since we have to change image naming convention due to (still
unresolved) bug with no permission to push the v2-1-test image
we've decided to change naming convention for all our cache
images to follow this - now available - "/" connvention to make
it better structured and easier to manage/understand.

Some more optimisations are implemented - Python, prod-build and
ci-manifest images are only pushed when "latest" image is prepared.
They are not needed for the COMMIT builds because we only need
final images for those builds. This simplified the code quite
a bit.

The push of cache image in CI is done in one job for both
CI and PROD images and the image is rebuilt again with
latest constraints, to account for the latest constraints
but to make sure that UPGRADE_TO_NEWER_DEPENDENCIES
is not set during the build (which invalidates the cache
for next non-upgrade builds)

Backwards-compatibility was implemented to allow PRs that have
not been upgraded to continue building after this one is merged,
also a workaround has been implemented to make this change
to work even if it is not merged yet to main.

This "legacy" mode will be removed in ~week when everybody rebase
on top of main.

Documentation is updated reflecting those changes.

potiuk commented on scripts/ci/libraries/_parallel.sh:
```diff
@@ -82,9 +85,9 @@ function parallel::monitor_loop() {
             fi
 
-            echo "${COLOR_BLUE}### The last lines for ${parallel_process} process: ${directory}/stdout ###${COLOR_RESET}"
+            echo "${COLOR_BLUE}### The last ${PARALLEL_TAIL_LENGTH} lines for ${parallel_process} process: ${directory}/stdout ###${COLOR_RESET}"
```
Needed to see both legacy and new image being pulled in logs

potiuk commented on .github/workflows/ci.yml:
```diff
@@ -1103,110 +1103,6 @@ ${{ hashFiles('.pre-commit-config.yaml') }}"
           path: /tmp/kind_logs_*
           retention-days: 7
 
-  push-prod-images-to-github-registry:
```
This is now merged into single job after constraints are pushed, and the build happens using those new constraints, so we are going to have new tested constraints always reflected in the latest cache images. 

I could also potentially separate out this change I think If you think it is too much in one PR.

potiuk commented on scripts/ci/libraries/_push_pull_remove_images.sh:
```diff
@@ -90,7 +72,7 @@ function push_pull_remove_images::check_and_rebuild_python_base_image_if_needed(
    dockerhub_python_version=$(docker run "${PYTHON_BASE_IMAGE}" python -c 'import sys; print(sys.version)')
    local local_python_version
-   local_python_version=$(docker run "${AIRFLOW_PYTHON_BASE_IMAGE}" python -c 'import sys; print(sys.version)')
+   local_python_version=$(docker run "${AIRFLOW_PYTHON_BASE_IMAGE}" python -c 'import sys; print(sys.version)' || true)
```
Accounts for case after `docker system prune --all`

potiuk commented on scripts/ci/libraries/_push_pull_remove_images.sh:
```diff
@@ -102,6 +84,10 @@ function push_pull_remove_images::check_and_rebuild_python_base_image_if_needed(
                 --label "org.opencontainers.image.source=https://github.com/${GITHUB_REPOSITORY}" \
                 -t "${AIRFLOW_PYTHON_BASE_IMAGE}" -
+  else
+      echo
```
Better diagnostics of what's going on

potiuk commented on scripts/ci/libraries/_push_pull_remove_images.sh:
```diff
@@ -151,8 +143,26 @@ function push_pull_remove_images::pull_ci_images_if_needed() {
+                echo "The CI image cache does not exist. This is likely the first time you build the image"
+                echo "Switching to 'local' cache for docker images"
+                echo
+                DOCKER_CACHE="local"
```
This is helpful if we have new python image not yet available in gitHubRegistry (next time - python 3.10)

ashb commented on IMAGES.rst:
```diff
@@ -246,19 +246,21 @@ Images with a commit SHA (built for pull requests and pushes)
-  ghcr.io/apache/airflow-<BRANCH>-pythonX.Y-build-v2:<COMMIT_SHA> - for production build stage
-  ghcr.io/apache/airflow-python-v2:X.Y-slim-buster-<COMMIT_SHA>   - for base Python images
+  ghcr.io/apache/airflow/<BRANCH>/ci/python<X.Y>:<COMMIT_SHA>         - for CI images
+  ghcr.io/apache/airflow/<BRANCH>/prod/python<X.Y>:<COMMIT_SHA>       - for production images
```
I'm glad they enabled this, it's so much nicer

potiuk replied:
indeed. Even if it's "sort of done" - only the first "/" is real separator - the others are just part of the name and translated to %2F in URL :D. The URLs how to get the packages look weird:

https://github.com/apache/airflow/pkgs/container/airflow%2Fv2-1-test%2Fprod-build%2Fpython3.8

potiuk replied:
The funny thing is that in Github UI tje first %2F in the name is treated as separator, even if it is part of the name :


![Screenshot from 2021-08-05 18-29-26](https://user-images.githubusercontent.com/595491/128387679-db18388f-b454-43ea-b149-5cdc851f1ec8.png)

ashb commented on scripts/ci/libraries/_parallel.sh:
```diff
@@ -82,9 +85,9 @@ function parallel::monitor_loop() {
+            echo "${COLOR_BLUE}### The last ${PARALLEL_TAIL_LENGTH} lines for ${parallel_process} process: ${directory}/stdout ###${COLOR_RESET}"
             echo
-            tail -2 "${directory}/stdout" || true
+            tail "-${PARALLEL_TAIL_LENGTH}" "${directory}/stdout" || true
```
`tail -s` is sleep. Didn't you mean `tail -n`?

potiuk replied:
No -s there :)

ashb replied:
.... I _swore_ I saw that as `-s${...`

👓

potiuk replied:
`tail -<NUMBER>` works same as `tail -n <NUMBER>`

ashb replied:
Yeah, I use that all the time. I just can't read.

ashb reviewed (approved):
One small comment about `tail` argument, otherwise LGTM.

Pre-emptively approving
